### DISCUSSION TITLE
There are cases where a result response is not received from iOS after completing the purchase.

### DISCUSSION THREAD

**Discussion by baesumin** [hyochan/react-native-iap#2661] [Q&A] [answered]
There is no problem on Android, but on iOS, the currentPurchase value may not come in from the useIAP hook after completing the purchase.

---
ex)
  const {
    getProducts,
    currentPurchase,
    currentPurchaseError,
    finishTransaction,
    requestPurchase,
  } = useIAP();
  
  
  useEffect(() => {
    if (currentPurchase) {
      console.log(currentPurchase); // -> Sometimes it doesn't work on iOS 
    }
  }, [currentPurchase]);

  ---

  Is there anything I'm missing?
  
  
  ---

Additionally, I found a situation where receipt details were not sent to the callback if user used another payment method instead of paying with a credit card.

**Comment by brent-charlton** [comment]
Does it behave the same in a none sandbox environment? I found the same issue and had to use the purchaseUpdatedListener to retrieve the purchase details instead of the hooks

**Comment by hyochan** [comment, marked as answer]
This is a **known issue in the iOS Sandbox environment**. The Sandbox environment can be unreliable and sometimes doesn't return purchase results properly.

As mentioned in the comments:
> "Does it behave the same in a none sandbox environment? I found the same issue and had to use the purchaseUpdatedListener to retrieve the purchase details instead of the hooks"

## Solutions

### 1. Use v14+ with onPurchaseSuccess callback (Recommended)

In v14+, use the `useIAP` hook with callbacks:

```typescript
import { useIAP } from 'react-native-iap';

const { requestPurchase, finishTransaction } = useIAP({
  onPurchaseSuccess: async (purchase) => {
    // This is more reliable than watching currentPurchase
    console.log('Purchase received:', purchase);
    await finishTransaction({ purchase, isConsumable: false });
  },
  onPurchaseError: (error) => {
    console.log('Purchase error:', error);
  },
});
```

### 2. Use purchaseUpdatedListener directly

If you need more control:

```typescript
import { purchaseUpdatedListener, finishTransaction } from 'react-native-iap';

useEffect(() => {
  const subscription = purchaseUpdatedListener(async (purchase) => {
    console.log('Purchase received:', purchase);
    // Process and finish transaction
  });

  return () => subscription.remove();
}, []);
```

### 3. Test in Production/TestFlight

Sandbox has known issues. Always verify behavior in TestFlight or production before concluding there's a bug.

## Note on Version

If you're using v12/v13, please upgrade to v14+ which has improved purchase handling:

- [Migration Guide (v13 to v14)](https://hyochan.github.io/react-native-iap/docs/guides/migration-v13-to-v14)
